opencog/atoms/distvalue/DistributionalValue.cc

```diff
@@ -0,0 +1,455 @@
+//Not recommended as it results in a DV with only 1 singleton set
+//which in some calculations is unusalbe as the chance of overlaps in the
+//Sets/Intervals is small
+DistributionalValue::DistributionalValue(double mode,double conf)
+	: Value(DISTRIBUTIONAL_VALUE)
+{
+	confidence_t cf = std::min(conf, 0.9999998);
+	double count = (DEFAULT_K * cf / (1.0 - cf));
+	//DV Interval with count 0 is undefined
+	count = std::max(count, 0.0000002);
+	DVKey k{Interval{mode}};
+	_value[k] = count;
+}
+
+DistributionalValuePtr DistributionalValue::createDV(const DVCounter &dvctr)
+{
+	return std::make_shared<const DistributionalValue>(dvctr);
+}
+
+DistributionalValuePtr DistributionalValue::createDV(double mode,
+                                                     double conf)
+{
+	return std::make_shared<const DistributionalValue>(mode,conf);
+}
+
+DistributionalValuePtr
+DistributionalValue::UniformDistributionalValue(const DVKey &k,int c)
+{
+	DVCounter dvctr;
+	dvctr[k] = c;
+	return createDV(dvctr);
+}
+
+
+DistributionalValuePtr
+DistributionalValue::UniformDistributionalValue(const DVKeySeq &ks,int c)
+{
+	DVCounter dvctr;
+	for (auto k : ks)
+	{
+		dvctr[k] = c;
+	}
+	return createDV(dvctr);
+}
+
+DistributionalValuePtr DistributionalValue::TRUE_TV()
+{
+	DVKey v1{Interval{1.0}};
+	DVKey v2{Interval{0.0}};
+	DVCounter dvc;
+	dvc[v1] = 1;
+	dvc[v2] = 0;
+	return std::make_shared<const DistributionalValue>(dvc);
+}
+DistributionalValuePtr DistributionalValue::FALSE_TV()
+{
+	DVKey v1{Interval{1.0}};
+	DVKey v2{Interval{0.0}};
+	DVCounter dvc;
+	dvc[v1] = 0;
+	dvc[v2] = 1;
+	return std::make_shared<const DistributionalValue>(dvc);
+}
+DistributionalValuePtr DistributionalValue::DEFAULT_TV()
+{
+	DVKey v1{Interval{1.0}};
+	DVKey v2{Interval{0.0}};
+	DVCounter dvc;
+	dvc[v1] = 0;
+	dvc[v2] = 0;
+	return std::make_shared<const DistributionalValue>(dvc);
+}
+
+bool DistributionalValue::is_uniform() const
+{
+	double val = _value.begin()->second;
+	for (auto p : _value)
+	{
+		if (val != p.second)
+		{
+			return false;
+		}
+	}
+	return true;
+}
+
+//Add Evidence for the provided key i.e increment the count of this Interval
+//Returns a new DVPtr with the update count
+DistributionalValuePtr DistributionalValue::add_evidence(const DVKey &h) const
+{
+	DVCounter newdvc = _value;
+	newdvc[h] += 1;
+	return createDV(newdvc);
+}
+
+//Combines this DV with the provided DV
+//and returns this new one ass a result
+DistributionalValuePtr DistributionalValue::merge(DistributionalValuePtr other) const
+{
+	DVCounter newdvc = _value;
+	newdvc += other->_value;
+	return createDV(newdvc);
+}
+
+//Flip all the counts
+DistributionalValuePtr DistributionalValue::negate() const
+{
+	double total = max_count() + min_count();
+	DVCounter res;
+	for (auto elem : _value)
+	{
+		res[elem.first] = total - elem.second;
+	}
+	return createDV(res);
+}
+
+//Get the lowest count of all Interals
+double DistributionalValue::min_count() const
```

Comment:
`min_count` and `max_count` need to be compbined into one, and return an `Interval` (not a `Bin`) 

Everyone who asks for min will always ask for max immediately afterwards, causing you to run the loop twice, paying all cache-line misses twice.  you can get both min and max "for free", because the additional branch will be almost completely seamped by cache-line effects.